### PR TITLE
🏷 Tagging feature tests 

### PR DESCRIPTION
**Pull request by Rabid-Dan**
Trello ticket https://trello.com/c/cDE58QW2/128-write-feature-tests-for-tagging

Adds features tests for the parts of submission tagging that have already been built, as well as one or two pending outlines for functionality still to be added.

Also adds a test helper file with a javascript function that allows for highlighting of text from within feature specs.

### CODE REVIEW

File: app/views/submissions/tag.html.erb [hunk 1]
```diff
@@ -9,7 +9,7 @@
     <h4 class="mb-4"><%= @submission.name %></h4>
     <div class="submission__taggable-text shadow p-3 js-taggable-submission-text"
          data-submission-id="<%= @submission.id %>">
-      <p><%= submission_text_with_tags(@submission.text, @submission_tags).html_safe %></p>
+      <p id="submissionText"><%= submission_text_with_tags(@submission.text, @submission_tags).html_safe %></p>
```
Comment (mermop) [inline app/views/submissions/tag.html.erb]: One thing I find useful is to prefix classes and IDs that are necessary for JS with `js-` - just useful to make it obvious that it'll have JS implications when reordering pages

File: spec/features/submissions/submissions_tag_spec.rb [hunk 2]
```diff
@@ -0,0 +1,59 @@
+require "rails_helper"
+
+RSpec.feature "Tagging a submission", js: true do
+  let!(:consultation) { FactoryBot.create(:consultation, :with_taxonomy_tags) }
+  let!(:submission) do
+    FactoryBot.create(:submission, :ready_to_tag, consultation: consultation,
+                                                  text: "cows dogs and sheep in the ocean riding waves")
```
Comment (mermop) [inline spec/features/submissions/submissions_tag_spec.rb]: Sounds rad 🐄🐕🐑🏄‍♀️

File: spec/features/submissions/submissions_tag_spec.rb [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+      it "can apply multiple tags to a piece of text" do
+        highlight_selection(5, 18)
+        find(".submission-tag[data-tag-name='#{tags.first.name}']").click
+        find(".submission-tag[data-tag-name='#{tags.second.name}']").click
+        expect(find(".tagged[data-tag-name='#{tags.first.name}']")).to have_text("dogs and sheep")
+        expect(find(".tagged[data-tag-name='#{tags.second.name}']")).to have_text("dogs and sheep")
+      end
+
+      xit "can apply tags that overlap"
```
Comment (mermop) [inline spec/features/submissions/submissions_tag_spec.rb]: 👍

File: spec/features/submissions/submissions_tag_spec.rb [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    end
+
+    xit "can remove tags that are applied to specific bits of text" do
+      highlight_selection(5, 18)
+      find(".submission-tag[data-tag-name='#{tags.first.name}']").click
+      expect(find(".tagged[data-tag-name='#{tags.first.name}']")).to have_text("dogs and sheep")
+      # still need functionality to remove tags in situ
+    end
+
+    it "refreshing the page displays existing tags" do
```
Comment (mermop) [inline spec/features/submissions/submissions_tag_spec.rb]: I would put this in its own context or describe block
```
context "when revisiting a submission with existing tags" do
  it "displays existing tags" do
...

File: spec/support/helper_modules/tagging_helpers.rb [hunk 5]
```diff
@@ -0,0 +1,47 @@
+module TaggingHelpers
+  # Takes the index of the first and last characters for selection and
+  # highlights them as you manually would with a mouse click and drag.
+  def highlight_selection(range_start, range_end)
```
Comment (mermop) [inline spec/support/helper_modules/tagging_helpers.rb]: Did you say you found that someone had had the same problem? Might be worth including a link to the stackoverflow answer etc if so

File: spec/support/helper_modules/tagging_helpers.rb [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+    page.execute_script(<<~JS, range_start, range_end)
+      (function(rangeStart, rangeEnd) {
```
Comment (mermop) [inline spec/support/helper_modules/tagging_helpers.rb]: A couple of code smells with this:

- This is a lot of JS to be in a Ruby method - that means it's not getting JS linters over it. Can you separate it out into a JS file?
- This is a lot of JS to be in a helper in general - it feels like it's changing the environment pretty significantly, in a way that means the tests might not be reflective of the real environment - did you look at any other ways to solve this problem? A different driver for example?

Comment (Rabid-Dan) [reply spec/support/helper_modules/tagging_helpers.rb]: I agree with both of those points but cant really find a way around them at the moment. The js is executed by a capybara function and I cant figure out a way of loading that from an external source. I also did a bunch of extra googling to see if I'd missed any other ways of doing the highlighting but that didn't turn up anything either.
It does feel pretty hacky doing it this way though so I'm more than happy to change this or wholesale remove it in the future if we can find a better way

**Pull request merged**